tests/src/Kernel/SmsFrameworkIncomingRouteTest.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace Drupal\Tests\sms\Kernel;
+
+use Symfony\Component\Routing\Exception\RouteNotFoundException;
+use Drupal\Core\Url;
+
+/**
+ * Tests incoming routes for gateway plugins.
+ *
+ * @group SMS Framework
+ */
+class SmsFrameworkIncomingRouteTest extends SmsFrameworkKernelBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static $modules = [
+    'system', 'sms', 'entity_test', 'user', 'field', 'telephone',
+    'dynamic_entity_reference', 'sms_test_gateway', 'basic_auth',
+  ];
+
+  /**
+   * The route provider.
+   *
+   * @var \Drupal\Core\Routing\RouteProviderInterface
+   */
+  protected $routeProvider;
+
+  /**
+   * The HTTP client.
+   *
+   * @var \GuzzleHttp\Client
+   */
+  protected $httpClient;
+
+  /**
+   * An incoming gateway instance.
+   *
+   * @var \Drupal\sms\Entity\SmsGatewayInterface
+   */
+  protected $incomingGateway;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+    $this->routeProvider = $this->container->get('router.route_provider');
+    $this->httpClient = $this->container->get('http_client');
+    $this->incomingGateway = $this->createMemoryGateway(['plugin' => 'incoming']);
```

Comment:
Same as above, was going to be used in multiple, but then those tests had to be moved to `SmsFrameworkIncomingBrowserTest`.

Will be created inline instead...